lib/bridge/QuitDebouncer.js

```diff
@@ -0,0 +1,125 @@
+    // This is to smooth incoming PART spam from IRC clients that suffer from a
+    // net-split (or other issues that lead to mass PART-ings)
+    this._debouncerForServer = {
+      // $server.domain: {
+            // $nick: {
+            //   promise: Promise, //Promise that resolves if the user joins a channel having quit
+            //   resolve: Function //function to call when a user joins
+            // }
+      // }
+    };
+
+    // Keep a track of the times at which debounceQuit was called, and use this to
+    // determine the rate at which quits are being received. This can then be used
+    // to detect net splits.
+    /*$server : [1477386173850, 1477386173825, ...]*/
+    this._quitTimesMs = {};
+    Object.keys(this.ircBridge.config.ircService.servers).forEach((domain) => {
+        this._quitTimesMs[domain] = [];
+    });
+}
+
+/**
+ * Called when the IrcHandler receives a JOIN. This resolves any promises to join that were made
+ * when a quit was debounced during a split.
+ * @param {string} nick The nick of the IRC user joining.
+ * @param {IrcServer} server The sending IRC server.
+ */
+QuitDebouncer.prototype.onJoin = function (nick, server) {
+    if (!this._debouncerForServer[server.domain]) {
+        return;
+    }
+    let rejoin = this._debouncerForServer[server.domain][nick];
+    if (rejoin) {
+        rejoin.resolve()
+    }
+}
+
+/**
+ * Debounce a QUIT received by the IrcHandler to prevent net-splits from spamming leave events
+ * into a room when incremental membership syncing is enabled.
+ * @param {Request} req The metadata request.
+ * @param {IrcServer} server The sending IRC server.
+ * @param {string} matrixUser The virtual user of the user that sent QUIT.
+ * @return {Promise} which resolves to true if a leave should be sent, false otherwise.
+ */
+QuitDebouncer.prototype.debounceQuit = Promise.coroutine(function*(req, server, matrixUser, nick) {
+    // Maintain the last windowMs worth of timestamps corresponding with calls to this function.
+    let now = Date.now();
+    let windowMs = 1000;// Window starts 1s ago
+    let threshold = 5;// Need 5Hz of quits to call net-split
+    this._quitTimesMs[server.domain].push(now);
+
+    // Filter out timestamps from more than 1s ago
+    this._quitTimesMs[server.domain] = this._quitTimesMs[server.domain].filter(
+        (t) => t > now - windowMs
+    );
+
+    // Wait for a short time to allow other potential splitters to send QUITs
+    yield Promise.delay(100);
+
+    let isSplitOccuring = this._quitTimesMs[server.domain].length > threshold;
+
+    if (this._debouncerForServer[server.domain] === undefined) {
+        this._debouncerForServer[server.domain] = {};
+    }
+
+    // TODO: This should be replaced with "disconnected" as per matrix-appservice-irc#222
+    let presence = "offline";
+    try {
+        yield this.ircBridge.getAppServiceBridge().getIntent(
+            matrixUser.getId()
+        ).client.setPresence(presence);
```

Comment:
Hmm, let's work this out.

Ignore presence for a moment. What happens when an IRC user quits the network? Each channel gets a QUIT, which the bridge translates into leave requests. That is the best behaviour because it most accurately describes what is going on in the membership list.

Now, how can presence changes improve this? Well for the channel case, it can't. For the PM case, it would be great to tell the user that the person they are talking to has disconnected, so "disconnected" would be great, and "offline" would be bad. They can't really leave the room because there is no one to invite them back into the room when they reconnect (the AS bot isn't a part of PM rooms). That's using "disconnected" outside the context of netsplits, and isn't one I previously considered.

Separately, we don't want to flood Synapse with leave requests. In netsplits, this `QuitDebouncer` will pick up on high rates of leaves and change the behaviour of the bridge: it will set the presence to "disconnected" rather than leave them from the Matrix room. Now my comment may make more sense: if we are not in a netsplit, we will leave the room. The Matrix user will never see the presence being set to disconnected because they won't be on the membership list anymore, so my question was "why should we bother then"?

The PM use case is a good counter-example of why we *should* bother.